File: 2026-03/2026-03-21/decode_qr.py
def decode_qr(qr_code):
    qr = qr_code[:]
    size = len(qr)
    orientation_markers_indices = {
        'bottom_left': [f'{size - 2}0', f'{size - 2}1', f'{size - 1}0', f'{size - 1}1'],
        'top_left': ['00', '01', '10', '11'],
        'top_right': [f'0{size - 2}', f'0{size - 1}', f'1{size - 2}', f'1{size - 1}']
    }
    orientation_markers_index_pairs = []
    data = ""

    for _, corner_indices in orientation_markers_indices.items():
        orientation_markers_index_pairs += corner_indices

    while not correct_orientation(qr, orientation_markers_index_pairs):
        rotated_qr = []

        for j in range(size):
            new_row = ""

            for i in range(size - 1, -1, -1):
                new_row += qr[i][j]

            rotated_qr.append(new_row)

        qr = rotated_qr[:]

    for i in range(size):
        for j in range(len(qr[i])):
            if f"{i}{j}" not in orientation_markers_index_pairs:
                data += qr[i][j]

    return data

def correct_orientation(qr, orientation_markers_index_pairs):
    for index_pair in orientation_markers_index_pairs:
        if qr[int(index_pair[0])][int(index_pair[1])] != "1":
            return False

    return True

# --- TEST SUITE ---

tests_text = r'''
1. decode_qr(["110011", "110011", "000000", "000000", "110000", "110001"]) should return "000000000000000000000001".
2. decode_qr(["100011", "000011", "000000", "000000", "110011", "110011"]) should return "000000000000000000000001".
3. decode_qr(["110011", "111111", "010000", "110000", "110011", "110100"]) should return "001101000011000000110100".
4. decode_qr(["011011", "101011", "101000", "100010", "110011", "111011"]) should return "010001000100010101010110".
5. decode_qr(["111100", "110001", "100011", "001101", "110011", "110011"]) should return "010000100100100101001110".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-21/decodeQr.js
function decodeQr(qrCode) {
  let qr = [...qrCode];
  const size = qr.length;
  const orientationMarkersIndices = {
    bottomLeft: [`${size - 2}0`, `${size - 2}1`, `${size - 1}0`, `${size - 1}1`],
    topLeft: ['00', '01', '10', '11'],
    topRight: [`0${size - 2}`, `0${size - 1}`, `1${size - 2}`, `1${size - 1}`]
  };
  let orientationMarkersIndexPairs = [];
  let data = "";

  for (const [_, cornerIndices] of Object.entries(orientationMarkersIndices))
    orientationMarkersIndexPairs = orientationMarkersIndexPairs.concat(cornerIndices);

  while(!correctOrientation(qr, orientationMarkersIndexPairs)) {
    const rotatedQr = [];

    for (let j = 0; j < size; ++j) {
      let newRow = "";

      for (let i = size - 1; i >= 0; --i)
        newRow += qr[i][j];

      rotatedQr.push(newRow);
    }

    qr = [...rotatedQr];
  }

  for (const i in qr)
    for (const j in qr[i])
      if (!orientationMarkersIndexPairs.includes(`${i}${j}`))
        data += qr[i][j];

  return data;
}

const correctOrientation = (qr, orientationMarkersIndexPairs) => {
  for (const indexPair of orientationMarkersIndexPairs)
    if (qr[indexPair[0]][indexPair[1]] !== "1")
      return false;

  return true;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. decodeQr(["110011", "110011", "000000", "000000", "110000", "110001"]) should return "000000000000000000000001".
2. decodeQr(["100011", "000011", "000000", "000000", "110011", "110011"]) should return "000000000000000000000001".
3. decodeQr(["110011", "111111", "010000", "110000", "110011", "110100"]) should return "001101000011000000110100".
4. decodeQr(["011011", "101011", "101000", "100010", "110011", "111011"]) should return "010001000100010101010110".
5. decodeQr(["111100", "110001", "100011", "001101", "110011", "110011"]) should return "010000100100100101001110".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);